### PR TITLE
Make ShardedTensor.size() equivalent to torch.Tensor.size() (#65087)

### LINKED ISSUE #65087: ShardedTensor.size() should be equivalent to torch.Tensor.size()
## 🚀 Feature
[ShardedTensor.size()](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/api.py#L575) should be equivalent to [torch.Tensor.size()](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/Functions.h#L193)

## Motivation
torch.Tensor.size() provides the following functionality:

```
import torch

t = torch.rand(3, 4)
print(t.size())
print(t.size(0))
print(t.size(1))
```

ShardedTensor.size() does not accept an optional dim to query the size of a particular dim.

## Pitch

Proposed API:

```
ShardedTensor.size(int dim = None) -> Union[torch.Size(), int]
```
```

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66012

Differential Revision: [D31345161](https://our.internmc.facebook.com/intern/diff/D31345161)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31345161).

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31345161).

**Comment by fduwjj** [comment]
Run python test/distributed/_sharded_tensor/test_sharded_tensor.py.

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31345161).

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31345161).

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31345161).

**Comment by fduwjj** [comment]
Added more unit tests, comments and exception check.

**Comment by fduwjj** [comment]
Fix the comment indentation and rebase.

**Review by pritamdamania87 — APPROVED** [review]
